plugins/kubernetes/app/models/watchers/base_watcher.rb

```diff
@@ -8,18 +8,24 @@ class BaseWatcher
     def initialize(watch_stream, log: true)
       @watch_stream, @log, = watch_stream, log
       @start_time = Time.now
+      @enabled = false
     end
 
     def start_watching(&block)
+      @original_thread_id = Thread.current.object_id
+      @enabled = true
       @watch_stream.each do |notice|
         handle_notice(notice, &block)
+        break if stop_watching?
       end
+      cleanup
     end
 
     def stop_watching
-      if @watch_stream
-        @watch_stream.finish
-        @watch_stream = nil
+      if @enabled
+        @enabled = false
+        # close the socket and stop receiving updates if called from a different thread than the original
+        cleanup if Thread.current.object_id != @original_thread_id
```

Comment:
@msufa just explained me the actual goal behind this fix, to allow us to stop watching the stream from a different Thread, if we find out that this Thread is stuck or something like that.

I misunderstood the whole point of this PR. I thought it was only to fix the crash we had.

Sorry @msufa. :-D